lib/puppet/agent.rb

```diff
@@ -1,6 +1,167 @@
+
+  private
+
+  {
+    'libcontract.so.1' => [
+      #function name,            return value type, parameter types, ...
+      [:ct_ctl_abandon,          Fiddle::TYPE_INT,  Fiddle::TYPE_INT],
+      [:ct_tmpl_activate,        Fiddle::TYPE_INT,  Fiddle::TYPE_INT],
+      [:ct_tmpl_clear,           Fiddle::TYPE_INT,  Fiddle::TYPE_INT],
+
+      [:ct_tmpl_set_informative, Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+      [:ct_tmpl_set_critical,    Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+      [:ct_pr_tmpl_set_param,    Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+      [:ct_pr_tmpl_set_fatal,    Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+
+      [:ct_status_read,          Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT, Fiddle::TYPE_VOIDP],
+
+      [:ct_status_get_id,        Fiddle::TYPE_INT,  Fiddle::TYPE_VOIDP],
+      [:ct_status_free,          Fiddle::TYPE_VOID, Fiddle::TYPE_VOIDP],
+    ],
+  }.each do |library, functions|
+    begin
+      libhandle = Fiddle::Handle.new(library)
+    rescue Fiddle::DLError => e
+      raise LoadError, e.message, e.backtrace
+    end
+
+    functions.each do |f|
+      define_method f[0], Fiddle::Function.new(libhandle[f[0].to_s], f[2..-1], f[1]).method(:call).to_proc
+    end
+  end
+
+  CTFS_PR_ROOT = File.join('', %w(system contract process))
+  CTFS_PR_TEMPLATE = File.join(CTFS_PR_ROOT, %q(template))
+  CTFS_PR_LATEST = File.join(CTFS_PR_ROOT, %q(latest))
+
+  CT_PR_PGRPONLY = 0x4
+  CT_PR_EV_HWERR = 0x20
+
+  CTD_COMMON = 0
+
+  def activate_new_contract_template()
+    begin
+      tmpl = File.open(CTFS_PR_TEMPLATE, File::RDWR)
+    rescue
+      Puppet.fatal 'Failed to open new contract template'
+    end
+
+    begin
+      tmpl_fd = tmpl.fileno
+
+      raise if ct_pr_tmpl_set_param(tmpl_fd, CT_PR_PGRPONLY) != 0
+      raise if ct_pr_tmpl_set_fatal(tmpl_fd, CT_PR_EV_HWERR) != 0
+      raise if ct_tmpl_set_critical(tmpl_fd, 0) != 0
+      raise if ct_tmpl_set_informative(tmpl_fd, CT_PR_EV_HWERR) != 0
+
+      raise if ct_tmpl_activate(tmpl_fd) != 0
+    rescue
+      tmpl.close()
+      Puppet.fatal 'Failed to configure / activate contract template'
+    end
+
+    @tmpl = tmpl
+  end
+
+  def deactivate_contract_template(where)
+    return 0 if @tmpl.nil?
+
+    tmpl = @tmpl
+    @tmpl = nil
+
+    begin
+      raise if ct_tmpl_clear(tmpl.fileno) != 0
+    rescue
+      Puppet.fatal "Failed to deactivate contract template in the #{where} process"
+    ensure
+      tmpl.close()
+    end
+  end
+
+  def get_latest_child_contract_id()
+    begin
+      stat = File.open(CTFS_PR_LATEST, File::RDONLY)
+    rescue
+      Puppet.fatal 'Unable to access latest contract'
+    end
+
+    begin
+      stathdl = Fiddle::Pointer.new(0)
+
+      raise if ct_status_read(stat.fileno, CTD_COMMON, stathdl.ref) != 0
+      ctid = ct_status_get_id(stathdl)
+      ct_status_free(stathdl)
```

Comment:
If  line 108 raises, then we never free the `stathdl`. A good way to handle this is to nest begin/rescues:

```ruby
stat = File.open(...)
begin
  stathdl = Fiddle::Pointer.new(0)
  begin
    raise if ct_status_read(stat.fileno, CTD_COMMON, stathdl.ref) != 0
    ...
  ensure
    ct_status_free(stathdl)
  end
rescue
  ...
ensure
  stat.close
end
```

Or create private helper methods to yield that param and ensure it's closed/released:

```ruby
def with_stat(&block)
  stat = File.open(...)
  yield stat
ensure
  stat.close
end
```

Then the caller can do:

```ruby
with_stat do |stat|
  with_fiddle_pointer do |stathdl|
    raise if ct_status_read(stat.fileno, CTD_COMMON, stathdl.ref) != 0
    ...
  end
end